datafusion/core/tests/sql/sql_api.rs

```diff
@@ -39,6 +52,99 @@ async fn unsupported_ddl_returns_error() {
     ctx.sql_with_options(sql, options).await.unwrap();
 }
 
+struct MockFunctionFactory;
+#[async_trait::async_trait]
+impl FunctionFactory for MockFunctionFactory {
+    #[doc = r" Crates and registers a function from [CreateFunction] statement"]
+    #[must_use]
+    #[allow(clippy::type_complexity, clippy::type_repetition_in_bounds)]
+    async fn create(
+        &self,
+        state: Arc<RwLock<SessionState>>,
+        statement: CreateFunction,
+    ) -> datafusion::error::Result<()> {
+        // this function is a mock for testing
+        // `CreateFunction` should be used to derive this function
+
+        let mock_add = Arc::new(|args: &[datafusion_expr::ColumnarValue]| {
+            let args = datafusion_expr::ColumnarValue::values_to_arrays(args)?;
+            let base =
+                datafusion_common::cast::as_float64_array(&args[0]).expect("cast failed");
+            let exponent =
+                datafusion_common::cast::as_float64_array(&args[1]).expect("cast failed");
+
+            let array = base
+                .iter()
+                .zip(exponent.iter())
+                .map(|(base, exponent)| match (base, exponent) {
+                    (Some(base), Some(exponent)) => Some(base.add_wrapping(exponent)),
+                    _ => None,
+                })
+                .collect::<arrow_array::Float64Array>();
+            Ok(datafusion_expr::ColumnarValue::from(
+                Arc::new(array) as arrow_array::ArrayRef
+            ))
+        });
+
+        let args = statement.args.unwrap();
+        let mock_udf = create_udf(
+            &statement.name,
+            vec![args[0].data_type.clone(), args[1].data_type.clone()],
+            Arc::new(statement.return_type.unwrap()),
+            datafusion_expr::Volatility::Immutable,
+            mock_add,
+        );
+
+        // we may need other infrastructure provided by state, for example:
+        // state.config().get_extension()
```

Comment:
🤔  maybe we should pass the config into `create` rather than `SessionContext`

Reply:
This comment is connected to previous, as getting `config` involves holding read lock. Read may not be as bad as write lock but still it will block writers. 

As `create` may take some time (like getting a ML model from repo) read lock may be problematic as well.

Function signature definitely looks better 

```rust
    async fn create(
        &self,
        state: &SessionConfig,
        statement: CreateFunction,
    ) -> Result<RegisterFunction>;
```

but lock is still an issue.